### ISSUE TITLE
.sismember no returning either promise or callback

### ISSUE THREAD

**Issue by lejoix** [luin/ioredis#248]
Hi,

I have yet another weird issue. The same part of code is handling sismember calls 
```
this.getConn().sismember(key, value)
            .then((reply) => {
                result = reply;
                resultReceived = true;
            }).catch((err) => {
                result = 0;
                resultReceived = true;
            });
```
``
this.getConn() simply returns the active ioredis instance.
``

The problem is that in most of the cases this works like a charm, in other cases it simply doesn't return anything, either with callback or promise, no error no catch no callback called.

another version of the same code
`this.getConn().sismember(key, value, (err, reply) => {
                result = reply;
                resultReceived = true;
)};`

The redis connection is active and it doesn't matter if the key exists or not (nor should it). Also redis monitor shows the command received.

I tried debugging the module, but got stuck after:

`line 524: redis.js - this.commandQueue.push({
      command: command,
      stream: stream,
      select: this.condition.select
    });`
 
Not sure what I can do next.

Thank you for your help.

**Comment by luin** [comment]
Never see this problem before. Is the problem reproducible? If it is, enable the debug mode (`DEBUG=ioredis:* node yourapp.js`) should give us more detailed information.

**Comment by lejoix** [comment]
Hi,

Thanks for the tip. So it's getting even weirder, before I saw it sent to redis (when I was debugging the module) but now it doesn't get sent and this is also consistent with the outpuyt of the debug (see below):

`ioredis:redis status[:3001]: [empty] -> connecting +2s
  ioredis:redis queue command[0] -> get(profile:123456789) +1ms
  ioredis:redis status[127.0.0.1:3001]: connecting -> connect +1ms
  ioredis:redis write command[0] -> info() +0ms
  ioredis:redis status[127.0.0.1:3001]: connect -> ready +1ms
  ioredis:connection send 1 commands in offline queue +0ms
  ioredis:redis write command[0] -> get(profile:123456789) +1ms
  ioredis:redis status[:3001]: [empty] -> connecting +2ms
  ioredis:redis queue command[0] -> sismember(all_playlist_slugs:123456789,asdasd) +0ms
  ioredis:redis status[127.0.0.1:3001]: connecting -> connect +1ms
  ioredis:redis write command[0] -> info() +0ms
`

**Comment by luin** [comment]
It seems there are two connections, one is ":3001" and the other is "127.0.0.1:3001". It's strange that the first connection didn't have the host part. Did you provide a host with an empty string or `null`?

`sismember` is sent via the ":3001" connection, and the following logs are the part related with ":3001":

```
ioredis:redis status[:3001]: [empty] -> connecting +2s
ioredis:redis status[:3001]: [empty] -> connecting +2ms
ioredis:redis queue command[0] -> sismember(all_playlist_slugs:123456789,asdasd) +0ms
ioredis:redis status[127.0.0.1:3001]: connecting -> connect +1ms
ioredis:redis write command[0] -> info() +0ms
```

`sismember` command will only be sent after the status going to "ready" ([empty] ----> connection ----> connect --`info`--> ready). So what the following several logs after `ioredis:redis write command[0] -> info() +0ms` you got? They should be similar to:

```
ioredis:redis write command[0] -> info() +0ms
ioredis:redis status[:3001]: connect -> ready +1ms
ioredis:connection send 1 commands in offline queue +0ms
ioredis:redis write command[0] -> sismember(all_playlist_slugs:123456789,asdasd) +0ms
```

**Comment by lejoix** [comment]
Ok I checked some more cleaned up. The weird error message was because I have a special case when starting node with env='test', but either way the problem is the same. I paste the new logs and then redis monitor logs. I have my own console logs also there

ioredis logs from node:
`Debugger listening on port 5858
  ioredis:redis status[:3001]: [empty] -> connecting +0ms
Loaded version: /v0 | Alias: stable | Default:true
PLY HTTP server listening on port 3333 in development mode
PLY HTTPS server listening on port 5433 in development mode
  ioredis:redis status[127.0.0.1:3001]: connecting -> connect +2s
  ioredis:redis write command[0] -> info() +1s
CONNECTED to redis on port :3001
  ioredis:redis status[127.0.0.1:3001]: connect -> ready +6ms
logged in true
  ioredis:redis write command[0] -> get(profile:566ae33524c12ab4786768b8) +408ms
  ioredis:redis write command[0] -> sismember(all_playlist_ids,59855178) +7s
`

redis monitor logs:
`1454946391.113656 [0 127.0.0.1:50440] "info"
1454946391.526532 [0 127.0.0.1:50440] "get" "profile:566ae33524c12ab4786768b8"
1454946398.954276 [0 127.0.0.1:50440] "sismember" "all_playlist_ids" "59855178"
`

So the command is queued and redis monitor catches it. But no response. Other commands work fine, also this command in other cases. Really weird

**Comment by luin** [comment]
Hmm... Seems that the log doesn't tell anything. A minimal reproducible example would be more helpful. Does the following code is able to reproduce the problem?

```javascript
var Redis = require('ioredis');
var redis = new Redis(50440);
redis.sismember('all_playlist_ids', '59855178', console.log.bind(console, 'got result'));
```

**Comment by luin** [comment]
Closed due to inactive status. Feel free to reopen it if the problem still exists.

**Comment by dophlin** [comment]
The problem still exists, sismember callback is never called. Any new solution?

**Comment by lejoix** [comment]
Sorry I didn't came back at this. I had to change my whole app's flow. My specific problem was that if I had multiple ioredis calls nested with callbacks, and there was a throw, I was doomed. So I had to rewrite my app to not follow this logic (handle errors differently). I would really recommend rewriting to Promises if you still are able to (I am not at this point, at least a few weeks of refactoring/bugfixing would be needed).
